_posts/2018-09-07-redis-ruby-bloom-filter.md

```diff
@@ -0,0 +1,355 @@
+      end
+    end
+
+    new_keys_count = keys.length.times.count { |i|
+      existing_indexes[i * bloom.hash_functions, bloom.hash_functions].include?(0)
+    }
+
+    total = redis.incrby(counter_key, new_keys_count)
+
+    if total == new_keys_count
+      redis.expire(filter_key, MAX_TTL.to_i)
+      redis.expire(counter_key, MAX_TTL.to_i)
+    end
+  end
+
+  def count
+    redis.get(counter_key).to_i
+  end
+
+  def include?(key)
+    values = redis.pipelined do
+      bloom.indexes_for(key).map { |index| redis.getbit(filter_key, index) }
+    end
+
+    !values.include?(0)
+  end
+
+  def field
+    redis.get(filter_key)
+  end
+
+  private
+
+  def redis
+    @redis ||= Redis.new
+  end
+
+  def bloom
+    @bloom ||= BloomFilter.new(@user, error_rate: 1)
+  end
+
+  def filter_key
+    @filter_key ||= "bloom:filter:#{key_suffix}"
+  end
+
+  def counter_key
+    @counter_key ||= "bloom:counter:#{key_suffix}"
+  end
+
+  def key_suffix
+    @key_suffix ||= bloom.fingerprint
+  end
+end
+```
+The `RedisBloomFilter` uses the `BloomFilter` implementation to produce the indexes that it just stores/retrieves them to/from Redis. It also implements a counter of how many unique elements are added to the filter by increasing the count only when it detects a unique insert. It uses redis pipelined that sends operations in batch to avoid latency and improve performance by about 5-6 times. It also sets a TTLs on the keys to expire them after a month and it exposes the field value to be consumed from the in-memory filter.
+
+Here's an example usage:
+
+
+```ruby
+redis_bloom_filter = RedisBloomFilter.new(user)
+
+redis_bloom_filter.insert(['user1@example.com', 'user2@example.com'])
+# => true
+
+redis_bloom_filter.count
+# => 2
+
+redis_bloom_filter.include?('user1@example.com')
+# => true
+
+redis_bloom_filter.include?('user2@example.com')
+# => true
+
+redis_bloom_filter.include?('user3@example.com')
+# => false
+```
+
+
+### In-memory Bloom filter
+
+With the Redis implementation we only solved half of the problem. We have a way to concurrently and fast add elements to the bloom filter in Redis, but we still need a way to check if a bloom filter could accept a given set of elements without actually inserting the elements in the filter. This is useful when we want to prevent a list import before importing the list or stop a campaign from sending before starting it.
+
+In order to achieve that, we need to implement an in-memory filter and [bitarray](https://github.com/peterc/bitarray) can help us with that. We have an important [PR](https://github.com/peterc/bitarray/pull/9) that changes the storage representation i.e. the bits order in bitarray to be as the Redis stores them internally and a way to intialize a bitarray with a given field. To test it, you can just fetch the `BitArray` that includes that patch from [here](https://github.com/dalibor/bitarray/blob/bits_order/lib/bitarray.rb).
```

Comment:
Grammar: `to be as the Redis stores them` -> `to match the way Redis stores them`

Reply:
Spelling: `intialize` -> `initialize`